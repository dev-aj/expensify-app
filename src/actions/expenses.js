import uuid from 'uuid';
import database from  '../firebase/firebase';
//- - - - - - - - - - - - --  - - -- - - -  --  -Action Generators(For Expense reducer) - - - - - - - - - - - - - - -
// - - - -- - -- -  What and how action generator work - -- - -
//1.Component calls action generator
//2.Action generator returns object
//3.Component dispatches object
//4.redux store changes
 
// - - - -- With asynchronous actions

//1.Component calls action generator
//2.Action generator returns function
//3.Component dispatches function(?)
//4.function runs (has the ability to dispatch other actions and do whatever it wants)


//Add Expenses

export const addExpense = (expense) => ({
        type: 'ADD_EXPENSE',
        expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) =>{
        const {
            description='',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense 
            }));
        });
    };
};

//Remove Expenses

export const removeExpense = ({ id } = {}) =>({
    type:'REMOVE_EXPENSE',
    id
});

//Edit Expense Action Generator

export const editExpense = ( id,update ) => ({
    type:'EDIT_EXPENSE',
    id,
    update
});

// - -- - -- - -- -- - To fetch data - - -- - -- 

//SET_EXPENSES 
export const setExpenses = (expenses) => ({
    type:'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];

            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id:childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    }
};