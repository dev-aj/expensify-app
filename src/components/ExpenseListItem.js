import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import NumberFormat from 'react-number-format';

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
    return (
        <div>
            <Link to={`/edit/${id}`}>
                <h3> {description} </h3>
            </Link>
            
            <p>
            <NumberFormat thousandSeparator={true} thousandsGroupStyle="lakh" prefix={'â‚¹'} displayType="text" value={amount} />
            - 
            {moment(createdAt).format("dddd, MMMM Do YYYY")}
            </p>
        </div>
    );
};


export default ExpenseListItem; 