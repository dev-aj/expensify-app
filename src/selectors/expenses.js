import moment from 'moment';


// - - - - - -  - - - -- - - - - ADDING FILTER LOGICS - -  - -- - - - - - - - - - - - - - 

//timestamps (milliseconds)
//0  = January1 1970(UNIX EPOCH)
//VAlid timestamps = 22121, 10, -8989

//Get Visible expenses

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day'):true
        const textMatch  = !expense.description || expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'date')
        return a.createdAt < b.createdAt ? 1 : -1;
        else
        return a.amount < b.amount ? 1 : -1;
    });
};

