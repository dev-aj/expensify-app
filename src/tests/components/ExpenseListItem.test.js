import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('Should render expenseListItem with expenses item', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
}); 