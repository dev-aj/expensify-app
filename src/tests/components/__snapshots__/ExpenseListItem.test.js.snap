// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expenseListItem with expenses item 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
       
      Gum
       
    </h3>
  </Link>
  <p>
    <NumberFormat
      allowEmptyFormatting={false}
      allowNegative={true}
      decimalSeparator="."
      displayType="text"
      fixedDecimalScale={false}
      getInputRef={[Function]}
      isAllowed={[Function]}
      isNumericString={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      onValueChange={[Function]}
      prefix="â‚¹"
      suffix=""
      thousandSeparator={true}
      thousandsGroupStyle="lakh"
      type="text"
      value={195}
    />
    -
    Thursday, January 1st 1970
  </p>
</div>
`;
